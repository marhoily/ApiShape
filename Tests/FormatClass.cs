using System;
using System.Collections.Generic;
using System.Net.NetworkInformation;
using ApiShape;
using ApprovalTests;
using ApprovalTests.Reporters;
using Xunit;
using static System.Reflection.BindingFlags;

namespace Tests
{
    public class C1 { }
    public sealed class C2 { }
    public abstract class C3 { }
    public sealed class C4 : C3 { }
    public class C5 : I1 { }
    public sealed class C6 : C5 { }
    public sealed class C7<T> { }
    public sealed class C8<T> where T : I1 { }
    public sealed class C9 : C5, I1 { }
    public static class C10 { }

    public interface I1 { }
    public interface I2 : I1 { }
    public interface I3<T> { }
    public interface I4<T> where T : I1 { }
    public interface I5 : I2 { }
    public interface I6<in T> { }
    public interface I7<out T> { }

    public struct S1 { }
    public struct S2 : I1 { }
    public struct S3 : I3<int> { }
    public struct S4<T> : I3<T> { }
    public struct S5<T> : I4<T> where T : I1 { }
    public struct S
    {
        public S(int a1) { F1 = 0; F2 = 0; P1 = 0; P2 = 0;P4 = 0; P5 = 0; }

        public const int X1 = 0;

        public int F1;
        public readonly int F2;
        public static int F4;

        public int P1 { get; }
        public int P2 { get; set; }
        public int P4 { get; private set; }
        public int P5 { private get; set; }
        public static int P13 { get; }

        public string this[int index] { get { return ""; } set { } }
        public string this[float index] { private get { return ""; } set { } }
        public string this[sbyte index] { get { return ""; } private set { } }
        public string this[double index] { set { } }
        public string this[byte index] => "";

        public void M1() { }
        public void M2(int a1) { }
        public void M3(int a1, int a2) { }
        public int M4() => 3;
        public int M5<T>() => 3;
        public int M6<T>() where T : I1 => 3;
        public static void M18() { }

        public static implicit operator S(int x) => new S();
        public static explicit operator int(S x) => 5;
        public static int operator  +(S x, int i) => 5;
    }

    public enum E1 { }
    public enum E2 : short { }
    public enum E3: byte { }
    public enum E4 : ulong { }
    [Flags] public enum E5 { }
    public enum E6 { V0, V1, V2 }
    [Flags] public enum E7
    {
        V0 = 0x1,
        V1 = 0x10,
        V2 = 0x100,
        V3 = 0x1000
    }

    public abstract class MM
    {
        private MM() { }
        private int F1;
        private const int X1 = 0;
        private string this[int index] => "";
        private void M27(params string[] arg) { }
        private event Action E5;

    }
    public abstract class M
    {
        public const int X1 = 0;
        public const string X2 = "blah";
        public const string X3 = null;
        public const double X4 = 0.0;
        public const double X5 = 15e9;

        public int F1;
        public readonly int F2;
        protected int F3;
        public static int F4;

        public int P1 { get; }
        public int P2 { get; set; }
        public int P3 { get; protected set; }
        public int P4 { get; private set; }
        public int P5 { private get; set; }
        public int P6 { protected get; set; }
        protected int P7 { get; }
        protected int P8 { get; set; }
        protected int P9 { private get; set; }
        protected int P10 { get; private set; }
        public abstract int P11 { get; }
        public virtual int P12 { get; }
        public static int P13 { get; }

        public string this[int index] { get { return ""; } set { } }
        protected string this[string index] { get { return ""; } set { } }
        public abstract string this[double index] { get; set; }
        public virtual string this[double index, int arg2] { get { return ""; } set { } }
        public abstract string this[byte i] { protected get; set; }
        public virtual string this[short i] { private get { return ""; } set { } }
        public virtual string this[ushort i] => "";

        protected M() { }
        public M(int a1) { }

        public abstract void M1();
        public abstract void M2(int a1);
        public abstract void M3(int a1, int a2);
        public abstract int M4();
        public abstract int M5<T>();
        public abstract int M6<T>() where T:I1;
        public virtual void M7() { }
        public void M8() { }

        protected abstract void M9();
        protected abstract void M10(int a1);
        protected abstract void M11(int a1, int a2);
        protected abstract int M12();
        protected abstract int M13<T>();
        protected abstract int M14<T>() where T:I1;
        protected virtual void M15() { }
        protected void M16() { }
        public abstract void M17(ref int a1, out int a2);
        public static void M18() { }
        public abstract int M19<T>() where T: class;
        public abstract int M20<T>() where T: struct;
        public abstract int M21<T>() where T: new();
        public abstract int M22<TLongLongLongLongLongArg1, TLongLongLongLongLongArg2, TLongLongLongLongLongArg3>();
        public abstract int M23<TLongLongLongLongLongArg1, TLongLongLongLongLongArg2, TLongLongLongLongLongArg3>(TLongLongLongLongLongArg1 a1, TLongLongLongLongLongArg2 a2, TLongLongLongLongLongArg3 a3);
        public abstract int M24(string longLongLongName, Dictionary<string, List<Uri>> anotherLongName, string notLongEnough);
        public abstract int M25(string arg = null);
        public abstract int M26(params string[] arg);

        public event EventHandler E1;
        public event EventHandler<NetworkAvailabilityEventArgs> E2 { add {} remove {} }
        public event Action E3;
        protected event Action E4;
    }

    public interface IM
    {
        int P1 { get; }
        int P2 { get; set; }
        int P3 { set; }

        void M1();
        void M2(int a1);
        void M3(int a1, int a2);
        int M4();
        int M5<T>();
        int M6<T>() where T : I1;
    }
    [UseReporter(typeof(VisualStudioReporter))]
    public sealed class FormatClass
    {
        [Fact]public void All() => Approvals.Verify(typeof(FormatClass).Assembly.GetShape());
    }
}