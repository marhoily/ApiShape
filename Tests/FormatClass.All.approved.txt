Full name: Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
Image runtime version: v4.0.30319
virtual class C1
{
    constructor();
}
abstract class C10
{
}
class C2
{
    constructor();
}
abstract class C3
{
}
class C4 : C3
{
    constructor();
}
virtual class C5
    I1
{
    constructor();
}
class C6 : C5
    I1
{
    constructor();
}
class C7<T>
{
    constructor();
}
class C8<T>
{
    constructor();
}
class C9 : C5
    I1
{
    constructor();
}
enum E1 : int
{
}
enum E2 : short
{
}
enum E3 : byte
{
}
enum E4 : ulong
{
}
[Flags]
enum E5 : int
{
}
enum E6 : int
{
    V0 = 0
    V1 = 1
    V2 = 2
}
[Flags]
enum E7 : int
{
    V0 = 1
    V1 = 16
    V2 = 256
    V3 = 4096
}
class FormatClass
{
    constructor();
    void All();
    void C1();
    void C10();
    void C2();
    void C3();
    void C4();
    void C5();
    void C6();
    void C7();
    void C8();
    void C9();
    void E1();
    void E2();
    void E3();
    void E4();
    void E5();
    void E6();
    void E7();
    void F1();
    void F2();
    void F3();
    void F4();
    void I1();
    void I2();
    void I3();
    void I4();
    void I5();
    void I6();
    void I7();
    void Im1();
    void Im2();
    void Im3();
    void Im4();
    void Im5();
    void Im6();
    void Ip1();
    void Ip2();
    void Ip3();
    void M1();
    void M10();
    void M11();
    void M12();
    void M13();
    void M14();
    void M15();
    void M16();
    void M17();
    void M2();
    void M3();
    void M4();
    void M5();
    void M6();
    void M7();
    void M8();
    void M9();
    void P1();
    void P10();
    void P2();
    void P3();
    void P4();
    void P5();
    void P6();
    void P7();
    void P8();
    void P9();
    void S1();
    void S2();
    void S3();
    void S4();
    void S5();
    void X1();
    void X2();
    void X3();
    void X4();
    void X5();
}
interface I1
{
}
interface I2
    I1
{
}
interface I3<T>
{
}
interface I4<T>
{
}
interface I5
    I1
    I2
{
}
interface I6<T>
{
}
interface I7<T>
{
}
interface IM
{
    int P1 { get; }
    int P2 { get; set; }
    int P3 { set; }
    virtual abstract void M1();
    virtual abstract void M2(int a1);
    virtual abstract void M3(int a1, int a2);
    virtual abstract int M4();
    virtual abstract int M5<T>();
    virtual abstract int M6<T : I1>();
}
abstract class M
{
    int F1;
    readonly int F2;
    static int F4;
    const int X1;
    const string X2;
    const string X3;
    const double X4;
    const double X5;
    string Item { get; set; }
    string Item { get; set; }
    string Item { get; set; }
    string Item { set; }
    string Item { set; }
    string Item { get; }
    int P1 { get; }
    int P11 { get; }
    int P12 { get; }
    int P13 { get; }
    int P2 { get; set; }
    int P3 { get; }
    int P4 { get; }
    int P5 { set; }
    int P6 { set; }
    constructor(int a1);
    virtual abstract void M1();
    virtual abstract void M17(int a1, out int a2);
    void M18();
    virtual abstract int M19<T>();
    virtual abstract void M2(int a1);
    virtual abstract int M20<T : ValueType>();
    virtual abstract int M21<T>();
    virtual abstract int M22<
        TLongLongLongLongLongArg1
        TLongLongLongLongLongArg2
        TLongLongLongLongLongArg3
    >();
    virtual abstract int M23<
        TLongLongLongLongLongArg1
        TLongLongLongLongLongArg2
        TLongLongLongLongLongArg3
    >(
        TLongLongLongLongLongArg1 a1
        TLongLongLongLongLongArg2 a2
        TLongLongLongLongLongArg3 a3
    );
    virtual abstract int M24(
        Dictionary<string, List<Uri>> anotherLongName
        string longLongLongName
        string notLongEnough
    );
    virtual abstract int M25(string arg = );
    virtual abstract int M26(string[] arg);
    virtual abstract void M3(int a1, int a2);
    virtual abstract int M4();
    virtual abstract int M5<T>();
    virtual abstract int M6<T : I1>();
    virtual void M7();
    void M8();
}
struct S1 : ValueType
{
}
struct S2 : ValueType
    I1
{
}
struct S3 : ValueType
    I3<int>
{
}
struct S4<T> : ValueType
    I3<T>
{
}
struct S5<T> : ValueType
    I4<T>
{
}
